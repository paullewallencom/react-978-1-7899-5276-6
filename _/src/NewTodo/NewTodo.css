.NewTodo {
  margin: 20px;
  padding: 20px;
  border: 2px solid #00ffd8;
  background: #ddffee;
  text-align: center;
}
