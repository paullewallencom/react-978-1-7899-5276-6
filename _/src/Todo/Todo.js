import React, { Component } from "react";
import { Button, ButtonGroup } from "reactstrap";
import styles from "./Todo.module.scss";

class Todo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      done: props.done,
      critical: props.critical
    };

    this.markAsDone = this.markAsDone.bind(this);
    this.removeTodo = this.removeTodo.bind(this);
    this.markCritical = this.markCritical.bind(this);
  }
  markCritical() {
    this.setState({ critical: true });
  }
  markAsDone() {
    this.setState({ done: true });
  }
  removeTodo() {
    this.props.removeTodo(this.props.id);
  }
  cssClasses() {
    let classes = [];
    if (this.state.critical) {
      classes = [styles.critical];
    } else {
      classes = [styles.todo];
    }
    if (this.state.done) {
      classes = [...classes, styles.done];
    }
    return classes.join(" ");
  }
  render() {
    return (
      <div className={this.cssClasses()}>
        {this.props.description}
        <br />
        <hr className={styles.hr} />
        <ButtonGroup>
          <Button
            className="MarkDone"
            onClick={this.markAsDone}
            color="success"
          >
            Mark as Done
          </Button>
          <Button
            className="RemoveTodo"
            onClick={this.removeTodo}
            color="warning"
          >
            Remove Me
          </Button>
          <Button
            className="MarkCritical"
            onClick={this.markCritical}
            color="danger"
          >
            Mark as Critical
          </Button>
        </ButtonGroup>
      </div>
    );
  }
}

export default Todo;
