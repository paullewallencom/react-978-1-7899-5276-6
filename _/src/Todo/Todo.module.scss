@import "../shared";

.todo {
  border: 2px solid black;
  text-align: center;
  background: $todo-normal;
  color: #333;
  margin: 20px;
  padding: 20px;
}

.done {
  background: $todo-complete;
}

.hr {
  border: 2px solid red;
}

.critical {
  composes: todo;
  background: $todo-critical;
}
