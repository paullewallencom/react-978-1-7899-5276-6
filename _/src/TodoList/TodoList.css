.TodoList {
  margin: 20px;
  padding: 20px;
  border: 2px solid #00d8ff;
  background: #ddeeff;
}
