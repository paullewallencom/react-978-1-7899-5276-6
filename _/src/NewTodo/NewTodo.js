import React, { Component } from "react";
import { Button, Input, InputGroup } from "reactstrap";
import "./NewTodo.css";

class NewTodo extends Component {
  constructor(props) {
    super(props);
    this.state = { item: "" };
    this.handleUpdate = this.handleUpdate.bind(this);
    this.addTodo = this.addTodo.bind(this);
  }
  addTodo() {
    this.props.addTodo(this.state.item);
    this.setState({ item: "" });
  }
  handleUpdate(event) {
    this.setState({ item: event.target.value });
  }
  render() {
    return (
      <div className="NewTodo">
        <InputGroup>
          <Input
            type="text"
            onChange={this.handleUpdate}
            value={this.state.item}
            placeholder="Input item name here..."
          />
          <Button onClick={this.addTodo} color="primary">
            Add
          </Button>
        </InputGroup>
      </div>
    );
  }
}

export default NewTodo;
