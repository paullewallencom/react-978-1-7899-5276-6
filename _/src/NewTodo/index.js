import NewTodo from "./NewTodo";
export default NewTodo;
