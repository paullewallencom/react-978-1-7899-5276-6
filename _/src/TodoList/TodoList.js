import React, { Fragment, Component } from "react";
import Todo from "../Todo/Todo";
import "./TodoList.css";

import NewTodo from "../NewTodo";
import Divider from "../Divider";

import { fetchTodos, createTodo, deleteTodo } from "../TodoService";

class TodoList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      items: [],
      loaded: false
    };

    this.addTodo = this.addTodo.bind(this);
    this.removeTodo = this.removeTodo.bind(this);
  }
  async componentDidMount() {
    const { todos } = await fetchTodos();
    this.setState({ items: todos, loaded: true });
  }
  async addTodo(description) {
    const { status } = await createTodo(description);
    if (status === 200) {
      const newItem = {
        id: this.state.items.length + 1,
        description: description,
        done: false,
        critical: false
      };
      this.setState({
        items: [...this.state.items, newItem]
      });
    }
  }
  async removeTodo(todoId) {
    const { status } = await deleteTodo(todoId);
    if (status === 200) {
      const filteredItems = this.state.items.filter(todo => {
        return todo.id !== todoId;
      });
      this.setState({ items: filteredItems });
    }
  }
  renderItems() {
    if (this.state.loaded) {
      return this.state.items.map(todo => (
        <Fragment key={"item-" + todo.description}>
          <Todo
            id={todo.id}
            key={todo.id}
            description={todo.description}
            removeTodo={this.removeTodo}
            done={todo.done}
            critical={todo.critical}
          />
          <Divider key={"divide-" + todo.description} />
        </Fragment>
      ));
    } else {
      return <p>Still Loading...</p>;
    }
  }
  render() {
    return (
      <div className="TodoList">
        <NewTodo addTodo={this.addTodo} />
        {this.renderItems()}
      </div>
    );
  }
}

export default TodoList;
