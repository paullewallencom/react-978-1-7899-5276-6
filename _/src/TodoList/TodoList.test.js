import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import TodoList from "./TodoList";
import NewTodo from "../NewTodo";
import Todo from "../Todo";

jest.mock("../TodoService", () => ({
  fetchTodos: jest.fn().mockReturnValue({ status: 200, todos: [] }),
  createTodo: jest.fn().mockReturnValue({ status: 200, todos: [] }),
  deleteTodo: jest.fn().mockReturnValue({ status: 200, todos: [] })
}));

describe(TodoList, () => {
  const component = shallow(<TodoList />);

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<TodoList />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders and matches our snapshot", () => {
    const component = renderer.create(<TodoList />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a TodoList component", () => {
    expect(component.contains(<div className="TodoList" />));
  });

  it("includes a NewTodo component", () => {
    expect(component.find(NewTodo)).toHaveLength(1);
  });

  it("renders the correct number of Todo components", () => {
    const todoCount = component.state("items").length;
    expect(component.find(Todo)).toHaveLength(todoCount);
  });

  it("adds another Todo when the addTodo function is called", async () => {
    const before = component.find(Todo).length;
    await component.instance().addTodo("New Item");
    component.update();
    const after = component.find(Todo).length;
    expect(after).toBeGreaterThan(before);
  });

  it("removes a Todo from the list when the remove Todo function is called", async () => {
    const before = component.find(Todo).length;
    const removeMe = component.state("items")[0];
    await component.instance().removeTodo(removeMe.id);
    component.update();
    const after = component.find(Todo).length;
    expect(after).toBeLessThan(before);
  });
});
