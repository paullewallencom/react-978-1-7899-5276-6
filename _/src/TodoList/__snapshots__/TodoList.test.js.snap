// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders and matches our snapshot 1`] = `
<div
  className="TodoList"
>
  <div
    className="NewTodo"
  >
    <div
      className="input-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Input item name here..."
        type="text"
        value=""
      />
      <button
        aria-label={null}
        className="btn btn-primary"
        onClick={[Function]}
        type="button"
      >
        Add
      </button>
    </div>
  </div>
  <p>
    Still Loading...
  </p>
</div>
`;
